planner/core/rule_aggregation_skew_rewrite.go

```diff
@@ -0,0 +1,262 @@
+					RetType:  newAggFunc.RetTp,
+				}
+			}
+			bottomAggSchema.Append(aggCol)
+
+			if newAggFunc.Name == ast.AggFuncCount {
+				cntIndexes = append(cntIndexes, i)
+				sumAggFunc, err := aggregation.NewAggFuncDesc(agg.ctx, ast.AggFuncSum,
+					[]expression.Expression{aggCol}, false)
+				if err != nil {
+					return nil
+				}
+				topAggFuncs = append(topAggFuncs, sumAggFunc)
+				topAggSchema.Columns[i] = &expression.Column{
+					UniqueID: agg.ctx.GetSessionVars().AllocPlanColumnID(),
+					RetType:  sumAggFunc.RetTp,
+				}
+			} else {
+				topAggFunc := aggFunc.Clone()
+				topAggFunc.Args = make([]expression.Expression, 0, len(aggFunc.Args))
+				topAggFunc.Args = append(topAggFunc.Args, aggCol)
+				topAggFuncs = append(topAggFuncs, topAggFunc)
+			}
+		}
+	}
+
+	// now create the bottom and top aggregate operators
+	bottomAgg := LogicalAggregation{
+		AggFuncs:     bottomAggFuncs,
+		GroupByItems: bottomAggGroupbyItems,
+		aggHints:     agg.aggHints,
+	}.Init(agg.ctx, agg.blockOffset)
+	bottomAgg.SetChildren(agg.children...)
+	bottomAgg.SetSchema(bottomAggSchema)
+
+	topAgg := LogicalAggregation{
+		AggFuncs:     topAggFuncs,
+		GroupByItems: agg.GroupByItems,
+		aggHints:     agg.aggHints,
+	}.Init(agg.ctx, agg.blockOffset)
+	topAgg.SetChildren(bottomAgg)
+	topAgg.SetSchema(topAggSchema)
+
+	if len(cntIndexes) == 0 {
+		appendSkewDistinctAggRewriteTraceStep(agg, topAgg, opt)
+		return topAgg
+	}
+
+	// it has count(), we have split it into sum()+count(), since sum() returns decimal
+	// we have to return a project operator that casts decimal to bigint
+	proj := LogicalProjection{
+		Exprs: make([]expression.Expression, 0, len(agg.AggFuncs)),
+	}.Init(agg.ctx, agg.blockOffset)
+	for _, column := range topAggSchema.Columns {
+		proj.Exprs = append(proj.Exprs, column)
+	}
+
+	// wrap sum() with cast function to keep output data type same
+	for _, index := range cntIndexes {
+		exprType := proj.Exprs[index].GetType()
+		targetType := agg.schema.Columns[index].GetType()
+		if !exprType.Equal(targetType) {
+			proj.Exprs[index] = expression.BuildCastFunction(agg.ctx, proj.Exprs[index], targetType)
+		}
+	}
+	proj.SetSchema(agg.schema.Clone())
+	proj.SetChildren(topAgg)
+	appendSkewDistinctAggRewriteTraceStep(agg, proj, opt)
+	return proj
+}
+
+func (a *skewDistinctAggRewriter) isQualifiedAgg(aggFunc *aggregation.AggFuncDesc) bool {
+	if aggFunc.Mode != aggregation.CompleteMode {
+		return false
+	}
+	if len(aggFunc.OrderByItems) > 0 || len(aggFunc.Args) > 1 {
+		return false
+	}
+
+	for _, arg := range aggFunc.Args {
+		if _, ok := arg.(*expression.Column); !ok {
+			if _, ok := arg.(*expression.Constant); !ok {
+				return false
+			}
+		}
+	}
+
+	switch aggFunc.Name {
+	case ast.AggFuncFirstRow, ast.AggFuncCount, ast.AggFuncSum, ast.AggFuncMax, ast.AggFuncMin:
+		return true
+	case ast.AggFuncBitAnd, ast.AggFuncBitOr, ast.AggFuncBitXor:
+		return true
+	case ast.AggFuncAvg:
+		return aggFunc.HasDistinct
+	default:
+		return false
+	}
```

Comment:
The PR's description says it only supports `count, avg, sum`. It's wrong here?

Reply:
Oh, the description needs to be updated.

Reply:
Please add some test cases for the other aggregation functions.

Reply:
It turns out bit agg funcs can't be pushed to TiFlash. I have added test cases.